### ISSUE TITLE
Does Civetweb support Brotli compression?

### ISSUE THREAD

**Issue by melwyndsouza1-eaton** [civetweb/civetweb#1305]
We are working on reducing our memory footprint. We thought of switching from gzip to brotli as we could see a good amount of difference for the compressed Web files. Is there a plan to include Brotli in future?

**Comment by bel2125** [comment]
Prior to your post I had no plan to add this compression.
Now I had a closer look and it seems possible, also from a license point of view.

As with gzip, there are two ways for compression: Either storing the already pre-compressed file, or supporting on the fly compression (or both). I need to have a closer look.

**Comment by RamPingle-eaton** [comment]
@bel2125 Hope you have some closer look as you said. Please let me know

**Comment by bel2125** [comment]
Yes, I had a look at "Brotli", but I'm not sure about this.
On Linux (at least Debian, Ubuntu, ...) there is a "libbrotli" that we might just use. For Windows there is no equivalent, as expected.

I was confused about the "dictionary" used by Brotli. Some site told me it's a compilation of words from English, Spanish, Chinese, Russian, Hindi, Arabic, while it doesn't work for Korean, Vietnamese, French, German, Japanese, Thai, ... and Brotli works best if you host the English Wikipedia. 

Currently I cannot estimate the effort for implementing this, in particular since I have not found how to work with this "dictionary" in practice. Are there any pages explaining this in detail?

**Comment by t-j-h** [comment]
OpenSSL 3.x has an interface to use Brotli ... including a way to dynamically pick up the support libraries on linux and windows without issues. See https://github.com/openssl/openssl/commit/12e96a23604a7aa1cd8f83486b02f1bcab6d468f for where it was added and the corresponding notes about the libraries that support it. It should save you a pile of time.

There is also the option of using the OpenSSL COMP interface to support it so you are one step removed from it. I don't know if this helps in the context of the original requester - i.e. if having an OpenSSL dependency is okay for them.

**Comment by SwissalpS** [comment]
Brötli is tied to secure connections, no need to bother with HTTP.

**Comment by bel2125** [comment]
Maybe to make it more clear what I mean:

There is "Content-Encoding":
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding
Here CivetWeb currently only uses gzip with zlib/libz.

And there is also compression inside TLS:
https://www.ietf.org/rfc/rfc3749.txt
CivetWeb doesn't do that in itself but OpenSSL can do that by itself (as far as I understood).

And finally there are some blog posts stating Chrome wants to support Brotli compression only for HTTPS connection, but the implementation actually works for HTTP and HTTPS, while Firefox indeed only supports it for HTTPS.
The more I read the more it becomes unclear what is actually to do (or I need to read a lot more to find what are up to date facts and what are old opinions of single authors).

**Comment by SwissalpS** [comment]
From what I heard, Google dropped the idea of supporting certain compressions for HTTP, not only brötli. If a compression becomes an attack vector when used without a secure layer, then it's not worth building support for non signed connections.

So I suggest you save some time, implement brötli for HTTPS and call it a day.
HTTP connections are getting rare anyway and that is a good thing.

**Comment by SwissalpS** [comment]
It 'works' to keep your cash stashed in your pillow, but it is better to at least put it in a fireproof safe ;)

**Comment by bel2125** [comment]
HTTP still has it's legitimate use, for example

in embedded devices - like my home weather station, accessible only from my home network. It's not exposed directly to the internet. The only "risk" is someone could be reading the outside temperature. It's not worth the trouble of certificate management, in particular since end customers with absolutely no IT skills must be able to install it. Also, for embedded devices without a clock certificates are problematic.

for applications communicating only locally, within the same trust domain. For example an application with the HTTP socket bound to localhost on a PC communicating to another application run by the same user.

Basically, do your thread/risk-analysis. HTTP isn't always a security issue, and HTTPS will not solve all possible security problems either.

Will those applications legitimately using HTTP **need** Brotli compression? Those embedded devices worked without as well and using compression for a localhost connection is probably even counterproductive. Maybe compression and HTTP do not have overlapping use cases.
